test/libsolidity/SolidityEndToEndTest.cpp

```diff
@@ -7382,6 +7382,33 @@ BOOST_AUTO_TEST_CASE(inline_assembly_function_access)
 	BOOST_CHECK(callContractFunction("x()") == encodeArgs(u256(10)));
 }
 
+BOOST_AUTO_TEST_CASE(inline_assembly_labels_with_stack_info)
+{
+	char const* sourceCode = R"(
+		contract C {
+			function f(uint y) {
+				assembly {
+					y 7
+					jump(fun)
+					exit[-1]:
+					y add
+					0 mstore
+					return(0, 0x20)
+					fun:
+					{
+						entry[a, b]:
+						a := div(a, b)
+						pop
```

Comment:
Right, we should probably make the functional / non-functional notation a bit more explicit, especially in the way arguments and return values are treated by opcodes. If an opcode takes its arguments from the implied stack (as opposed to from a functional expression argument) or if it leaves data on the stack (which is not consumed as part of the same statement), a `!` has to be appended - otherwise an error is generated. If a `!` is present, but it does not satisfy the above requirements, an error is also generated. So the following is fine:
```
store(1, 2)
1 2 sstore!
let x := sload(4)
sload!(4) =: x
```
While these are not:
```
store!(1, 2)
1 2 sstore
let x := sload!(4)
sload(4) =: x
```